Boost 1.61 hides fcontext

Which means Recursor cannot use Boost Context on 1.61 and up.

ph1's hack does not work for me on osx with brew boost 1.61.0:
```
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:8:7: error: no member named 'make_fcontext' in namespace 'boost::context'; did you mean 'boost::context::detail::make_fcontext'?
using boost::context::make_fcontext;
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      boost::context::detail::make_fcontext
/usr/local/include/boost/context/fcontext.hpp:33:12: note: 'boost::context::detail::make_fcontext' declared here
fcontext_t make_fcontext( void * sp, std::size_t size, void (* fn)( transfer_t) );
           ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:39:7: error: no member named 'fcontext_t' in namespace 'boost::context'; did you mean 'boost::context::detail::fcontext_t'?
using boost::context::fcontext_t;
      ^~~~~~~~~~~~~~~~~~~~~~~~~~
      boost::context::detail::fcontext_t
/usr/local/include/boost/context/fcontext.hpp:23:15: note: 'boost::context::detail::fcontext_t' declared here
typedef void* fcontext_t;
              ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:40:7: error: no member named 'jump_fcontext' in namespace 'boost::context'; did you mean 'boost::context::detail::jump_fcontext'?
using boost::context::jump_fcontext;
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      boost::context::detail::jump_fcontext
/usr/local/include/boost/context/fcontext.hpp:31:12: note: 'boost::context::detail::jump_fcontext' declared here
transfer_t jump_fcontext( fcontext_t const to, void * vp);
           ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:71:5: error: no matching function for call to 'jump_fcontext'
    jump_fcontext (reinterpret_cast<fcontext_t*>(&ctx->uc_mcontext),
    ^~~~~~~~~~~~~
/usr/local/include/boost/context/fcontext.hpp:31:12: note: candidate function not viable: requires 2 arguments, but 3 were provided
transfer_t jump_fcontext( fcontext_t const to, void * vp);
           ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:84:5: error: no matching function for call to 'jump_fcontext'
    jump_fcontext (reinterpret_cast<fcontext_t*>(&ctx->uc_mcontext),
    ^~~~~~~~~~~~~
/usr/local/include/boost/context/fcontext.hpp:31:12: note: candidate function not viable: requires 2 arguments, but 3 were provided
transfer_t jump_fcontext( fcontext_t const to, void * vp);
           ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:107:9: error: no matching function for call to 'jump_fcontext'
    if (jump_fcontext (reinterpret_cast<fcontext_t*>(&octx.uc_mcontext),
        ^~~~~~~~~~~~~
/usr/local/include/boost/context/fcontext.hpp:31:12: note: candidate function not viable: requires 2 arguments, but 3 were provided
transfer_t jump_fcontext( fcontext_t const to, void * vp);
           ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:119:23: error: no matching function for call to 'make_fcontext'
    ctx.uc_mcontext = make_fcontext (&ctx.uc_stack[ctx.uc_stack.size()],
                      ^~~~~~~~~~~~~
/usr/local/include/boost/context/fcontext.hpp:33:12: note: candidate function not viable: no known conversion from 'void (*)(const intptr_t)' to 'void (*)(boost::context::detail::transfer_t)' for 3rd argument
fcontext_t make_fcontext( void * sp, std::size_t size, void (* fn)( transfer_t) );
           ^
In file included from mtasker_context.cc:6:
./mtasker_fcontext.cc:124:5: error: no matching function for call to 'jump_fcontext'
    jump_fcontext (reinterpret_cast<fcontext_t*>(&args.prev_ctx),
    ^~~~~~~~~~~~~
/usr/local/include/boost/context/fcontext.hpp:31:12: note: candidate function not viable: requires 2 arguments, but 3 were provided
transfer_t jump_fcontext( fcontext_t const to, void * vp);
           ^
8 errors generated.
make[2]: *** [mtasker_context.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

--- a/pdns/mtasker_fcontext.cc
+++ b/pdns/mtasker_fcontext.cc
@@ -23,7 +23,11 @@
 #include <exception>
 #include <cassert>
 #include <type_traits>
+#if BOOST_VERSION > 106100
+#include <boost/context/detail/fcontext.hpp>
+#else
 #include <boost/context/fcontext.hpp>
+#endif
 #include <boost/version.hpp>
 
 using boost::context::make_fcontext;

--- a/pdns/recursordist/configure.ac
+++ b/pdns/recursordist/configure.ac
@@ -54,9 +54,15 @@ AC_DEFUN([PDNS_SELECT_CONTEXT_IMPL], [
       LDFLAGS="$LDFLAGS $BOOST_THREAD_LDFLAGS"
     fi
     AC_MSG_NOTICE([checking whether the Boost context library actually links...])
-    BOOST_FIND_HEADER([boost/context/fcontext.hpp], [ : ], [
-      BOOST_FIND_LIB([context], [$1], [boost/context/fcontext.hpp], [[]])
-    ])
+    if test $boost_major_version -ge 161; then
+      BOOST_FIND_HEADER([boost/context/detail/fcontext.hpp], [ : ], [
+        BOOST_FIND_LIB([context], [$1], [boost/context/detail/fcontext.hpp], [[]])
+      ])
+    else
+      BOOST_FIND_HEADER([boost/context/fcontext.hpp], [ : ], [
+        BOOST_FIND_LIB([context], [$1], [boost/context/fcontext.hpp], [[]])
+      ])
+    fi
     case $boost_cv_lib_context in
       (yes)
         AC_MSG_NOTICE([MTasker will use the Boost context library for context switching])

--- a/pdns/recursordist/configure.ac
+++ b/pdns/recursordist/configure.ac
@@ -42,6 +42,7 @@ PDNS_CHECK_NETWORK_LIBS
 
 # Boost Context was introduced in 1.51 (Aug 2012), but there was an immediate
 # API break in 1.52 (Nov 2012), so we only support that, and later.
+pdns_context_library="System V ucontexts"
 
 AC_DEFUN([PDNS_SELECT_CONTEXT_IMPL], [
   AC_MSG_CHECKING([whether Boost is new enough to use the context library...])
@@ -65,6 +66,7 @@ AC_DEFUN([PDNS_SELECT_CONTEXT_IMPL], [
     fi
     case $boost_cv_lib_context in
       (yes)
+        pdns_context_library="Boost context"
         AC_MSG_NOTICE([MTasker will use the Boost context library for context switching])
         ;;
       *)
@@ -206,4 +208,5 @@ AS_IF([test "x$systemd" != "xn"],
   [AC_MSG_NOTICE([systemd: yes])],
   [AC_MSG_NOTICE([systemd: no])]
 )
+AC_MSG_NOTICE([Context library: $pdns_context_library])
 AC_MSG_NOTICE([])

